### PR TITLE
Do not encode response when the status indicates empty content

### PR DESCRIPTION
**Pull request by minwoox**
…mpty

Motivation:
`EncodingService` should not try to encode the body when the status indicates that the content it empty.

Modifications:
- Check the status if it indicates the content is empty in `EncodingService`.
- Downgrade Micrometer to 1.3.6 becuase 1.4.0 is Non-LTS.

Result:
- `EncodingService` does not encode the content when the status is 204, 205 and 304

### CODE REVIEW

**Comment by anuraaga** [comment]
I'm wondering is this an optimization or a bug. What happens when encoding an empty response right now, I guess it's a no-op?

**Comment by minwoox** [comment]
> I'm wondering is this an optimization or a bug. What happens when encoding an empty response right now, I guess it's a no-op?

It's a bug. 😄 the `EncodingSerivce` emits the gzipped emtpy response which is `void`.

**Comment by trustin** [comment]
> I guess it's a no-op?

Not a no op - It emits an empty yet valid gzip stream.

File: core/src/main/java/com/linecorp/armeria/server/HttpResponseSubscriber.java [hunk 1]
```diff
@@ -180,12 +182,19 @@ public void onNext(HttpObject o) {
...
             case NEEDS_TRAILERS: {
-                if (o instanceof HttpData || o instanceof ResponseHeaders) {
+                if (o instanceof ResponseHeaders) {
                     failAndRespond(new IllegalStateException(
-                            "published an HttpData or a ResponseHeaders: " + o +
+                            "published a ResponseHeaders: " + o +
                             " (expected: an HTTP trailers). service: " + service()));
                     return;
                 }
+                if (o instanceof HttpData) {
```
Comment (anuraaga) [inline core/src/main/java/com/linecorp/armeria/server/HttpResponseSubscriber.java]: Just wondering is this change enough to fix the issue?

File: core/src/main/java/com/linecorp/armeria/server/encoding/HttpEncodedResponse.java [hunk 2]
```diff
@@ -76,8 +76,8 @@ protected HttpObject filter(HttpObject obj) {
             final ResponseHeaders headers = (ResponseHeaders) obj;
 
             // Skip informational headers.
-            final String status = headers.get(HttpHeaderNames.STATUS);
-            if (ArmeriaHttpUtil.isInformational(status)) {
+            final HttpStatus status = headers.status();
+            if (status.isInformational() || status.isContentAlwaysEmpty()) {
```
Comment (anuraaga) [inline core/src/main/java/com/linecorp/armeria/server/encoding/HttpEncodedResponse.java]: Can we move this check to `shouldEncodeResponse`?

File: core/src/test/java/com/linecorp/armeria/server/HttpResponseSubscriberTest.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+public class HttpResponseSubscriberTest {
+
+    @RegisterExtension
+    static final ServerExtension server = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) throws Exception {
+            sb.service("/", (ctx, req) -> {
+                final ResponseHeaders headers = ResponseHeaders.builder(HttpStatus.NO_CONTENT).contentType(
+                        MediaType.PLAIN_TEXT_UTF_8).build();
+                // Add CONTINUE not to validate when creating HttpResponse.
```
Comment (anuraaga) [inline core/src/test/java/com/linecorp/armeria/server/HttpResponseSubscriberTest.java]: `// HTTP does not allow a payload with a CONTINUE status, make sure we ignore it.`

File: core/src/test/java/com/linecorp/armeria/server/HttpResponseSubscriberTest.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+                return HttpResponse.of(ResponseHeaders.of(HttpStatus.CONTINUE), headers,
+                                       HttpData.ofUtf8("foo"));
+            });
+        }
+    };
+
+    @Test
+    void httpResponseSubscriberDoNotThrowExceptionWhenContentIsNotEmpty() {
```
Comment (anuraaga) [inline core/src/test/java/com/linecorp/armeria/server/HttpResponseSubscriberTest.java]: DoesNot

File: core/src/main/java/com/linecorp/armeria/server/HttpResponseSubscriber.java (hunk 1 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/server/HttpResponseSubscriber.java]: I think so. Instead of resetting, we can just ignore the content and call `subscription.request(1)` so that the response is completed in `onComplete`.

File: core/src/test/java/com/linecorp/armeria/server/HttpResponseSubscriberTest.java (hunk 3 above)
Comment (minwoox) [reply core/src/test/java/com/linecorp/armeria/server/HttpResponseSubscriberTest.java]: Actually, I added `CONTINUE` to call `static HttpResponse of(HttpObject... objs) {...}`, let me rephrase a little bit.

Comment (minwoox) [reply core/src/test/java/com/linecorp/armeria/server/HttpResponseSubscriberTest.java]: Removed the comment and use `streaming()` 😉

File: core/src/main/java/com/linecorp/armeria/server/encoding/HttpEncodedResponse.java (hunk 2 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/server/encoding/HttpEncodedResponse.java]: Ah thanks. It should be.

**Comment by trustin** [comment]
Thanks, @minwoox !

**Comment by minwoox** [comment]
Thank you all!
